### ISSUE TITLE
before_type_cast not work when it called by an object through association mapping and index

### ISSUE THREAD

**Issue by jiso0jung** [rails/rails#44549]
### Steps to reproduce
```ruby
# db/schema.rb
ActiveRecord::Schema[7.0].define(version: 2022_02_25_061737) do
  create_table "boxes", force: :cascade do |t|
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
  end

  create_table "chocolates", force: :cascade do |t|
    t.string "flavor"
    t.integer "box_id", null: false
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
    t.index ["box_id"], name: "index_chocolates_on_box_id"
  end

  add_foreign_key "chocolates", "boxes"
end
```
```ruby
# app/models/box.rb
class Box < ApplicationRecord
  has_many :chocolates, dependent: :destroy

  def got_chocolate(flavor)
    chocolates.create!(flavor:)
  end
end
```
```ruby
# app/models/chocolate.rb
class Chocolate < ApplicationRecord
  enum flavor: { matcha: '녹차', strawberry: '딸기' }
  belongs_to :box
end
```
``` ruby
# spec/models/chocolate_spec.rb
require 'rails_helper'

RSpec.describe Chocolate, type: :model do
  let(:box) { create(:box) }

  before do
    box.got_chocolate(Chocolate.flavors[:matcha])
    box.got_chocolate(Chocolate.flavors[:strawberry])
  end

  it do
    chocolates = box.chocolates
    p chocolates.first.flavor_before_type_cast
    p chocolates[0].flavor_before_type_cast

    chocolates = Chocolate.where(box:)
    p chocolates.first.flavor_before_type_cast
    p chocolates[0].flavor_before_type_cast

    should belong_to(:box)
  end
end
```

### Expected behavior
"녹차" is the value and "matcha" is the key of the enum hash

```ruby
"녹차"
"녹차"
"녹차"
"녹차"
```

### Actual behavior
```ruby
"녹차"
"matcha"      # ????
"녹차"
"녹차"
```

### System configuration
**Rails version**: 7.0.2

**Ruby version**: 3.1.0

### repository
https://github.com/jiso0jung/sample-proj

**Comment by faraazahmad** [comment]
Why does only the 2nd case fail out of the 4?

**Comment by hahmed** [comment]
Actually looking at this again, the repo script below works for english;

```
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "rails", "7.0.2"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :boxes do |t|
  end

  create_table :chocolates do |t|
    t.string :flavor
    t.integer :box_id, null: false
    t.index [:box_id], name: "index_chocolates_on_box_id"
  end

  add_foreign_key :chocolates, :boxes
end

class Box < ActiveRecord::Base
  has_many :chocolates, dependent: :destroy

  def got_chocolate(flavor)
    chocolates.create!(flavor: flavor)
  end
end

class Chocolate < ActiveRecord::Base
  # enum flavor: { matcha: "녹차", strawberry: "딸기" }
  enum flavor: { matcha: "matcha", strawberry: "strawberry" }
  belongs_to :box
end

class BugTest < Minitest::Test
  def test_create_box_with_chocolate
    box = Box.create!
    box.got_chocolate(Chocolate.flavors[:matcha])

    all_chocolates = box.chocolates
    chocolate = all_chocolates.first
    other_chocolate = all_chocolates[0]

    assert_equal chocolate, other_chocolate
    assert_equal chocolate.flavor_before_type_cast, other_chocolate.flavor_before_type_cast
  end
end
```

If we switch to the korean enum values, we get an encoding error;

```
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "rails", "7.0.2"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :boxes do |t|
  end

  create_table :chocolates do |t|
    t.string :flavor
    t.integer :box_id, null: false
    t.index [:box_id], name: "index_chocolates_on_box_id"
  end

  add_foreign_key :chocolates, :boxes
end

class Box < ActiveRecord::Base
  has_many :chocolates, dependent: :destroy

  def got_chocolate(flavor)
    chocolates.create!(flavor: flavor)
  end
end

class Chocolate < ActiveRecord::Base
  enum flavor: { matcha: "녹차", strawberry: "딸기" }
  # enum flavor: { matcha: "matcha", strawberry: "strawberry" }
  belongs_to :box
end

class BugTest < Minitest::Test
  def test_create_box_with_chocolate
    box = Box.create!
    box.got_chocolate(Chocolate.flavors[:matcha])

    all_chocolates = box.chocolates
    chocolate = all_chocolates.first
    other_chocolate = all_chocolates[0]

    assert_equal chocolate, other_chocolate
    assert_equal chocolate.flavor_before_type_cast, other_chocolate.flavor_before_type_cast
  end
end
# =>
# Failure:
# BugTest#test_create_box_with_chocolate [issue_44549.rb:64]:
# --- expected
# +++ actual
# @@ -1 +1,3 @@
# -"녹차"
# +# encoding: US-ASCII
# +#    valid: true
# +"matcha"
```

**Comment by jiso0jung** [comment]
@hahmed  Does it work if the value has different string literal from key of the enum? In my case, it does not work when the value has different alphabet string literal value from key name. ex: { matcha: "MATCHA" }

**Comment by hahmed** [comment]
@jiso0jung correct, it does not work if the value is different from the key.

Let me update the repro scripts.

**Comment by nashby** [comment]
As far as I know that's expected behavior. You can find more info here https://github.com/rails/rails/pull/40456

TL; DR:
You should use `chocolate.flavor_for_database` to get proper DB value.

**Comment by jiso0jung** [comment]
@nashby Oh I checked it work. Thank you. But there's still a few questions.
On below code, number 1,2,3,4 reference same record in table, and they return same value except number 2.
Why `before_type_cast` cannot reference value only in number 2?

```ruby
# spec/models/chocolate_spec.rb
require 'rails_helper'

RSpec.describe Chocolate, type: :model do
  let(:box) { create(:box) }

  before do
    box.got_chocolate(Chocolate.flavors[:matcha])
    box.got_chocolate(Chocolate.flavors[:strawberry])
  end

  it do
    chocolates = box.chocolates
    p chocolates.first.flavor_before_type_cast #1
    p chocolates[0].flavor_before_type_cast   #2

    chocolates = Chocolate.where(box:)
    p chocolates.first.flavor_before_type_cast #3
    p chocolates[0].flavor_before_type_cast    #4   

    should belong_to(:box)
  end
end
```
